analyzers/src/SonarAnalyzer.CSharp/Extensions/MethodDeclarationSyntaxExtensions.cs

```diff
@@ -17,13 +17,55 @@
  * along with this program; if not, write to the Free Software Foundation,
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
+using StyleCop.Analyzers.Lightup;
 
 namespace SonarAnalyzer.Extensions
 {
     internal static class MethodDeclarationSyntaxExtensions
     {
+        /// <summary>
+        /// Returns true if the method throws exceptions or returns null.
+        /// </summary>
+        public static bool ThrowsOrReturnsNull(this MethodDeclarationSyntax syntaxNode) =>
+            syntaxNode.DescendantNodes().OfType<ThrowStatementSyntax>().Any() ||
+            syntaxNode.DescendantNodes().OfType<ExpressionSyntax>().Any(expression => expression.IsKind(SyntaxKindEx.ThrowExpression)) ||
+            syntaxNode.DescendantNodes().OfType<ReturnStatementSyntax>().Any(returnStatement => returnStatement.Expression.IsKind(SyntaxKind.NullLiteralExpression)) ||
+            // For simplicity this returns true for any method witch contains a NullLiteralExpression but this could be a source of FNs
+            syntaxNode.DescendantNodes().OfType<ExpressionSyntax>().Any(expression => expression.IsKind(SyntaxKind.NullLiteralExpression));
+
+        public static bool IsExtensionMethod(this BaseMethodDeclarationSyntax methodDeclaration) =>
+            methodDeclaration.ParameterList.Parameters.Count > 0
+            && methodDeclaration.ParameterList.Parameters[0].Modifiers.Any(s => s.ValueText == "this");
+
         public static bool HasReturnTypeVoid(this MethodDeclarationSyntax methodDeclaration) =>
             methodDeclaration.ReturnType.ToString() == "void";
```

Comment:
@martin-strecker-sonarsource  Regarding the proposal to use `methodDeclaration.ReturnType is PredefinedTypeSyntax { TypeKind: SyntaxKind.VoidKeyword}`, the `PredefinedTypeSyntax` or its ancestors does not seem to have `TypeKind` property (or anything similar, see [docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.predefinedtypesyntax?view=roslyn-dotnet-4.2.0)).

Reply:
Should be something like that `methodDeclaration.ReturnType is PredefinedTypeSyntax { Keyword: { RawKind: (int)SyntaxKind.VoidKeyword } }`